x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java

```diff
@@ -96,25 +100,21 @@ public static class Builder extends FieldMapper.Builder {
         private final IndexSettings indexSettings;
 
         private final Parameter<String> inferenceId = Parameter.stringParam(
-            "inference_id",
+            INFERENCE_ID_FIELD,
             false,
             mapper -> ((SemanticTextFieldType) mapper.fieldType()).inferenceId,
             null
```

Comment:
Should we use the `defaultValue` param here to provide the defaults? I think that plays nicely with the mapping serialization (so we avoid to serialize default param values), and we may avoid some checks further down the road.

Reply:
We can use `defaultValue` for `inference_id`, but we can't for `search_inference_id`. This is because the default value for `search_inference_id` is dependent on the value of `inference_id`:

If `inference_id` == `DEFAULT_INFERENCE_ID`, `search_inference_id` default should be `DEFAULT_SEARCH_INFERENCE_ID`

If `inference_id` != `DEFAULT_INFERENCE_ID`, `search_inference_id` default should be null so that if `search_inference_id` is not specified, the inference endpoint specified by `inference_id` is used at ingest and search time.

Reply:
> If inference_id != DEFAULT_INFERENCE_ID, search_inference_id default should be null so that if search_inference_id is not specified, the inference endpoint specified by inference_id is used at ingest and search time.

I see - maybe you could use a Supplier for that, which is [what stringParam does](https://github.com/elastic/elasticsearch/blob/af7ed9515f24a2dfbab552877a1bafedf99938e8/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L1087) under the hood. 

As mentioned, my concern is not really the implementation (though if it can be simplified with existing constructs, that's better), but that the serialization is done correctly.

Reply:
Thanks for pointing this out! I added a couple methods to `FieldMapper` to expose creating a string parameter with a default value supplier.